### DISCUSSION TITLE
Define a single schema with end-to-end Type safety

### DISCUSSION THREAD

**Discussion by airtonix** [Thinkmill/keystatic#211] [Q&A] [answered]
Currently in a NetlifyCMS project, I'm defining the CMS UI via typescript and then making the content available as imports via [contentlayer](https://www.contentlayer.dev/), which generates types for all my generated content.

Problem with this approach is that I end up defining the schema twice: 
- one for the UI
- one for the generated data and how I use it in the api endpoints and things like `getStaticPaths` , `getStaticProps` and `getServerSideProps`.

Will KeyStatic do this for me?  

edit: to clarify 

Here's a basic level example https://github.com/officialrajdeepsingh/contentLayerNetlifycms/blob/main/contentlayer.config.ts

- This is the content layer Post schema: https://github.com/officialrajdeepsingh/contentLayerNetlifycms/blob/main/contentlayer.config.ts#L4-L57, it ends up generating :
  - [types ](https://github.com/officialrajdeepsingh/contentLayerNetlifycms/blob/main/.contentlayer/generated/types.d.ts#L12-L30)
  - [importable content](https://github.com/officialrajdeepsingh/contentLayerNetlifycms/blob/main/.contentlayer/generated/index.mjs#L7)

- This would be the related (but functionally separate) UI schema for the cms: https://github.com/officialrajdeepsingh/contentLayerNetlifycms/blob/main/public/admin/config.yml#L7-L21
  - this chap has done what most others mention in their tutorials: define it in yaml in the public folder
  - but you can define it via typescript.
  - if you want to use custom cms widgets, then instantiation via the js/ts api is required. although you can probably still import the yaml instead of defining all that in ts. Obviously as your cms use case scales you'd want to move to typescript. 

I'm hoping that I wont need to : 
- use content-layer
- define two different schemas

And I hope that I will
- have programmatic access to my content (along with generated types)

**Comment by simonswiss** [comment]
Hey!

Sorry about the delay in this answer. 

**Update:** I got excited with this, and created a demo repo where I re-create a content structure mirroring what you've got in your links, to give you an idea of how everything works.

The TL;DR of how you can pull schema-defined types in the frontend is here: https://github.com/simonswiss/keystatic-typesafety-demo/blob/main/src/app/(app)/page.tsx#L9-L15

---

I feel like [Keystatic's Reader API](https://thinkmill.notion.site/Reader-API-4f0cffe819e44037bb576bb095ccf014) does all you want, and means you don't need to define your schema/types in two different places.

The schema you define in your config file will generate Types you can bring in your frontend templates.

The Reader API gives you a mechanism to read data from your file system, and also import the Types generated from the schema definition.

Check out this `Entry` type here:

https://github.com/Thinkmill/keystatic/blob/ecd5eda2a7f8275d4a8bfec00bb800ff575d74e9/packages/keystatic/src/reader/index.ts#L45

With it, you can get the type for a collection entry like so:

`Entry<typeof myConfig['collections']['myCollection']>`

Hope this helps!

**Reply by airtonix** [reply]
This seems to be a fairly simple approach, but one thing that will surely occur in projects that need to scale is that the common parts of the scheme will be composable.

This means there will be a need to create types from those composable fragments.

seems like we can do it like this: 

<details><summary>`keystatic.config.ts`</summary>
<p>


```ts
import { config, fields, collection, Collection } from '@keystatic/core';
import { Entry } from '@keystatic/core/reader';

const FakeKeystaticCollectionNoopString = 'dont care about this field' as const;

const PublishablePageSchema = {
    title: fields.slug({ name: { label: 'Title' }}),
    date: fields.date({ label: 'date' }),
    stage: fields.text({ label: 'Status' }),
}
export type PublishablePage = Entry<Collection<
    typeof PublishablePageSchema,
    typeof FakeKeystaticCollectionNoopString
>>;

const IconifyingPageSchema = {
    icon: fields.text({ label: 'Icon' }),
}
export type IconifyingPage = Entry<Collection<
    typeof IconifyingPageSchema,
    typeof FakeKeystaticCollectionNoopString
>>;

const TaggablePageSchema = {
    tags: fields.array(
        fields.text({ label: 'tag' }),
        {
            label: 'Tag',
            itemLabel: props => props.value
        }
    ),
}
export type TaggablePage = Entry<Collection<
    typeof TaggablePageSchema,
    typeof FakeKeystaticCollectionNoopString
>>;

const WithCommentsPageSchema = {
    comments: fields.checkbox({ label: 'Allow comments' }),
}
export type WithCommentsPage = Entry<Collection<
    typeof WithCommentsPageSchema,
    typeof FakeKeystaticCollectionNoopString
>>;

const TemplatablePageSchema = {
    theme: fields.text({ label: 'Theme' }),
    template: fields.select({
        label: 'Template',
        options: [{
            label: 'Post',
            value: 'post'
        },{
            label: 'Page',
            value: 'page'
        }],
        defaultValue: 'post',
    }),
}
export type TemplatablePage = Entry<Collection<
    typeof TemplatablePageSchema,
    typeof FakeKeystaticCollectionNoopString
>>;

const WithBannerPageSchema = {
    img: fields.conditional(
        fields.checkbox({ label: 'use banner image' }),
        {
            false: fields.empty(),
            true: fields.object({
                url: fields.url({ label: 'banner url'}),
                credit: fields.text({ label: 'credit' })
            })
        }
    ),
}
export type WithBannerPage = Entry<Collection<
    typeof WithBannerPageSchema,
    typeof FakeKeystaticCollectionNoopString
>>;

const ContentfulPageSchema = {
    content: fields.document({
        label: 'Content',
        formatting: true,
        dividers: true,
        links: true,
        images: true,
    }),
}
export type ContentfulPage = Entry<Collection<
    typeof ContentfulPageSchema,
    typeof FakeKeystaticCollectionNoopString
>>;


export default config({
    storage: {
        kind: 'local',
    },
    collections: {
        pages: collection({
            label: 'About',
            slugField: 'title',
            path: 'src/content/about/*',
            format: { contentField: 'content' },
            entryLayout: 'content',
            schema: {
                ...PublishablePageSchema,
                ...TemplatablePageSchema,
                ...ContentfulPageSchema,
            }
        }),
        posts: collection({
            label: 'Posts',
            slugField: 'title',
            path: 'src/content/posts/*',
            format: { contentField: 'content' },
            entryLayout: 'content',
            schema: {
                ...PublishablePageSchema,
                ...WithBannerPageSchema,
                ...IconifyingPageSchema,
                ...TaggablePageSchema,
                ...TemplatablePageSchema,
                ...WithCommentsPageSchema,
                ...ContentfulPageSchema
            },
        }),
    },
});
```

</p>
</details> 



<details><summary>`@/services/KeyStatic`</summary>
<p>

```ts
import { createReader } from '@keystatic/core/reader';
import keystaticConfig from '../../keystatic.config';

export const ContentStore = createReader(process.cwd(), keystaticConfig);
```

</p>
</details> 

<details><summary>`@/templates/BlogDetailPage`</summary>
<p>


```ts
import React, { PropsWithChildren } from 'react'

import { ArticlePage } from '@/components/ArticlePage'
import { BrandingBlockContainer } from '@/containers/BrandingBlockContainer'
import type { WithBannerPage } from '../../keystatic.config'

export const BlogDetailPage = ({
  title,
  date,
  tags,
  img,
  children,
}: PropsWithChildren<{
  title: string
  date: string,
  tags: string[],
}& WithBannerPage>) => {

  return (
      <ArticlePage
        title={title}
        date={date}
        tags={tags}
        branding={BrandingBlockContainer}
      >
        {children}
      </ArticlePage>
  )
}

```

</p>
</details> 


<details><summary>`@/app/b/[slug]/page.tsx`</summary>
<p>


```tsx
import { ContentStore } from '@/services/Keystatic';
import { BlogDetailPage } from '@/templates/post';
import { DocumentRenderer } from '@keystatic/core/renderer';


export default async function Post({ params }: {
    params: { slug: string }
}) {

  const post = await ContentStore.collections.posts.read(params.slug);

  if (!post) {
    return (
        <h1>Not found</h1>
    )
  }
  
  return (
    <>
      <BlogDetailPage
        date={post.date}
        img={post.img}
        >
        <DocumentRenderer document={await post.content()} />
      </BlogDetailPage>
    </>
  );
}
```

</p>
</details>

**Reply by airtonix** [reply]
actually i don't think this solves the problem.

i only get types that makes sense in the context of the admin ui, not the resolved content.

**Reply by airtonix** [reply]
related:

1. https://github.com/Thinkmill/keystatic/issues/855#issue-2077719867
2. https://github.com/Thinkmill/keystatic/issues/854

**Reply by airtonix** [reply]
@simonswiss sorry the reader doesn't give me what i need.

**Reply by airtonix** [reply, marked as answer]
So i started picking apart the types behind the read api and it looks like we need this type exposed: 

```ts
type ValueForReadingWithMode<
    Schema extends ComponentSchema,
    ResolveLinkedFiles extends boolean | undefined
> = ResolveLinkedFiles extends true
    ? ValueForReadingDeep<Schema>
    : ValueForReading<Schema>;

type ResolvedSchema<Schema extends Record<string, ComponentSchema>, SlugField = void> = {
    [Key in keyof Schema]: SlugField extends Key
        ? Schema[Key] extends SlugFormField<any, any, any, infer SlugSerializedValue>
            ? SlugSerializedValue
            : ValueForReadingWithMode<Schema[Key], true>
        : ValueForReadingWithMode<Schema[Key], true>;
}
```

[Screencast from 2024-01-12 17-26-23.webm](https://github.com/Thinkmill/keystatic/assets/61225/92a9fbcb-2b87-47c9-b81f-dfd6c71bb825)


```ts
// keystatic.config.ts
import { config, fields, collection } from '@keystatic/core';
import type { Singleton, ComponentSchema, ValueForReadingDeep, ValueForReading, SlugFormField } from '@keystatic/core';
import type { Entry } from '@keystatic/core/reader';


type ValueForReadingWithMode<
    Schema extends ComponentSchema,
    ResolveLinkedFiles extends boolean | undefined
> = ResolveLinkedFiles extends true
    ? ValueForReadingDeep<Schema>
    : ValueForReading<Schema>;

type ResolvedSchema<Schema extends Record<string, ComponentSchema>, SlugField = void> = {
    [Key in keyof Schema]: SlugField extends Key
        ? Schema[Key] extends SlugFormField<any, any, any, infer SlugSerializedValue>
            ? SlugSerializedValue
            : ValueForReadingWithMode<Schema[Key], true>
        : ValueForReadingWithMode<Schema[Key], true>;
}

const PublishablePageSchema = {
    title: fields.slug({ name: { label: 'Title' }}),
    date: fields.date({ label: 'date' }),
    stage: fields.text({ label: 'Status' }),
}

type Foo = Entry<Singleton<typeof PublishablePageSchema>>
export type PublishablePage = ResolvedSchema<
    typeof PublishablePageSchema, 'title'
>

const IconifyingPageSchema = {
    icon: fields.text({ label: 'Icon' }),
}

type Bar = Entry<Singleton<typeof IconifyingPageSchema>>
export type IconifyingPage = ResolvedSchema<
    typeof IconifyingPageSchema
>
```
